docs/namespace_config_spec.md

```diff
@@ -0,0 +1,309 @@
+# The "Keto Schema Language" Specification
+
+Enforcing fine-grained permissions is a critical building block of mature
+technology solutions that protect privacy and identity in the information age. A
+number of proprietary languages to represent permission already exists, namely
+the [Authzed Schema Language](https://docs.authzed.com/reference/schema-lang)
+and [Auth0 FGA](https://docs.fga.dev/). Many permissions are defined by normal
+developers who typically are most familiar with web technologies like JavaScript
+or Typescript. We therefore see the need for a developer-friendly configuration
+language for permissions that has such a small learning curve that it can be
+picked up, used, and understood with confidence by most developers. We therefore
+chose to define our permissions configuration language as a subset of the most
+common general-purpose programming language: JavaScript/TypeScript.
+
+The Keto Schema Language is a syntactical subset of TypeScript. Along with type
+definitions for the syntax elements of the language (such as `Namespace` or
+`Context`), users can get context help from their IDE while writing the
+configuration.
+
+## Notation
+
+The syntax is specified using Extended Backus-Naur Form (EBNF):
+
+```ebnf
+Production  = production_name "=" [ Expression ] "." .
+Expression  = Alternative { "|" Alternative } .
+Alternative = Term { Term } .
+Term        = production_name | token [ "…" token ] | Group | Option | Repetition .
+Group       = "(" Expression ")" .
+Option      = "[" Expression "]" .
+Repetition  = "{" Expression "}" .
+```
+
+Productions are expressions constructed from terms and the following operators,
+in increasing precedence:
+
+```ebnf
+|   alternation
+()  grouping
+[]  option (0 or 1 times)
+{}  repetition (0 to n times)
+```
+
+Lower-case production names are used to identify lexical tokens. Non-terminals
+are in CamelCase. Lexical tokens are enclosed in double quotes "".
+
+The form `a … b` represents the set of characters from a through b as
+alternatives. The horizontal ellipsis `…` is also used elsewhere in the spec to
+informally denote various enumerations or code snippets that are not further
+specified.
+
+## Configuraton text representation
+
+The configuration is encoded in UTF-8.
+
+## Lexical elements
+
+### Comments
+
+1. Line comments start with the character sequence // and stop at the end of the
+   line.
+2. General comments start with the character sequence /* and stop with the first
+   subsequent character sequence */.
+3. Documentation comments start with the character sequence /** and stop with
+   the first subsequent character sequence */.
+
+### Identifiers
+
+Identifiers name program entities such as variables and types. An identifier is
+a sequence of one or more letters and digits. The first character in an
+identifier must be a letter.
+
+```ebnf
+identifier = letter { letter | digit } .
+digit      = "0" … "9"
+letter     = "A" … "Z" | "a" … "z" | "-" | "_"
```

Comment:
Are identifiers things like variables? If so, can we really use `-`?

```
// This would not compile in TS:
var foo-bar
```

Reply:
Good point, let's stick with the [JS syntax for identifiers](https://developer.mozilla.org/en-US/docs/Glossary/Identifier): unicode letters, $, _, and digits (0-9).

Reply:
Yes, I noticed that as well. Thanks!